File: apps/backend/billingms/src/main/java/com/example/billingms/domain/model/Invoice.java
package com.example.billingms.domain.model;

import java.time.Instant;
import java.util.List;

/**
 * 精算書（US21・[ADR-027] 決定 3・決定 4）。
 *
 * <p><strong>算出中の精算書は存在しない</strong>（決定 3）。経理担当者が確定操作をした時点で
 * 初めて発行される。下書きを持つと、下書きのまま忘れられた精算書が溜まる——それを見つける
 * 手段をまた作ることになる。
 *
 * <p><strong>発行した精算書の金額は動かない</strong>（決定 4）。請求書は荷主へ出す約束であり、
 * 出したあとに黙って変わると請求の根拠が消える。金額を変える手段をここに置かない。
 * 訂正は US23（IT12）で「取り消して出し直す」形にする。
 *
 * <p><strong>根拠を持ったまま発行する。</strong>基本料金の内訳（区間数・重量・貨物種別）、
 * 割引率、キャンセル料の料率と申請時の状態——金額だけを持つと、あとから
 * 「なぜその金額か」を答えられない。経理担当者は請求の根拠を荷主に説明する。
 */
public final class Invoice {

    private final InvoiceId invoiceId;
    private final BillingBookingId cargoBookingId;
    private final BillingShipperId shipperId;
    private final TransportCharge charge;
    private final DiscountPolicy discountPolicy;
    private final List<InvoiceLineItem> lineItems;
    private final CancellationFee cancellationFee;
    private final TaxRate taxRate;
    private final PaymentStatus paymentStatus;
    private final Instant issuedAt;

    private Invoice(InvoiceId invoiceId, BillingBookingId cargoBookingId,
            BillingShipperId shipperId, TransportCharge charge, DiscountPolicy discountPolicy,
            List<InvoiceLineItem> lineItems, CancellationFee cancellationFee, TaxRate taxRate,
            PaymentStatus paymentStatus, Instant issuedAt) {
        this.invoiceId = invoiceId;
        this.cargoBookingId = cargoBookingId;
        this.shipperId = shipperId;
        this.charge = charge;
        this.discountPolicy = discountPolicy;
        // **写して持つ。** 呼び出し元が渡したあとの書き換えでこちらの中身が変わらないように
        this.lineItems = List.copyOf(lineItems);
        this.cancellationFee = cancellationFee;
        this.taxRate = taxRate;
        this.paymentStatus = paymentStatus;
        this.issuedAt = issuedAt;
    }

    /**
     * 精算書を発行する（US21-4・US21-5）。
     *
     * <p><strong>発行の時点では未入金である</strong>（決定 3）。入金の確認は US23。
     */
    public static Invoice issue(InvoiceId invoiceId, BillingBookingId cargoBookingId,
            BillingShipperId shipperId, TransportCharge charge, DiscountPolicy discountPolicy,
            List<InvoiceLineItem> lineItems, CancellationFee cancellationFee, TaxRate taxRate,
            Instant issuedAt) {
        if (invoiceId == null) {
            throw new IllegalArgumentException("請求番号を指定してください");
        }
        if (cargoBookingId == null) {
            throw new IllegalArgumentException("予約を指定してください");
        }
        if (shipperId == null) {
            throw new IllegalArgumentException("荷主を指定してください");
        }
        if (charge == null) {
            throw new IllegalArgumentException("基本料金の根拠を指定してください");
        }
        if (discountPolicy == null) {
            throw new IllegalArgumentException("割引方針を指定してください");
        }
        if (taxRate == null) {
            throw new IllegalArgumentException("税率を指定してください");
        }
        if (issuedAt == null) {
            throw new IllegalArgumentException("発行日時を指定してください");
        }
        return new Invoice(invoiceId, cargoBookingId, shipperId, charge, discountPolicy,
                lineItems == null ? List.of() : lineItems, cancellationFee, taxRate,
                PaymentStatus.PENDING, issuedAt);
    }

    /**
     * 永続化された行から復元する。
     *
     * <p><strong>ここでは検査しない</strong>（新しい不変条件は既存行を壊す）。
     * 検査するのは新規に受け入れるとき（{@link #issue}）である。
     */
    public static Invoice restore(InvoiceId invoiceId, BillingBookingId cargoBookingId,
            BillingShipperId shipperId, TransportCharge charge, DiscountPolicy discountPolicy,
            List<InvoiceLineItem> lineItems, CancellationFee cancellationFee, TaxRate taxRate,
            PaymentStatus paymentStatus, Instant issuedAt) {
        return new Invoice(invoiceId, cargoBookingId, shipperId, charge, discountPolicy,
                lineItems == null ? List.of() : lineItems, cancellationFee, taxRate,
                paymentStatus, issuedAt);
    }

    public InvoiceId invoiceId() {
        return invoiceId;
    }

    public BillingBookingId cargoBookingId() {
        return cargoBookingId;
    }

    public BillingShipperId shipperId() {
        return shipperId;
    }

    /** 基本料金の根拠（決定 1）。 */
    public TransportCharge charge() {
        return charge;
    }

    public Money baseAmount() {
        return charge.baseAmount();
    }

    /** 割引率。**割引が無ければ {@code null}**——0% と契約なしを区別する（[ADR-012]）。 */
    public DiscountRate discountRate() {
        return discountPolicy.rate();
    }

    public Money discountAmount() {
        return discountPolicy.discountOf(baseAmount());
    }

    /** キャンセル料。キャンセルされていなければ {@code null}。 */
    public CancellationFee cancellationFee() {
        return cancellationFee;
    }

    /** 調整の明細。**発行後は足せない**（決定 4）。 */
    public List<InvoiceLineItem> lineItems() {
        return lineItems;
    }

    public TaxRate taxRate() {
        return taxRate;
    }

    /** 税を含まない小計。 */
    public Money subtotal() {
        Money amount = baseAmount().subtract(discountAmount());
        if (cancellationFee != null) {
            amount = amount.add(cancellationFee.amount());
        }
        for (InvoiceLineItem item : lineItems) {
            amount = amount.add(item.amount());
        }
        return amount;
    }

    public Money taxAmount() {
        return taxRate.taxOf(subtotal());
    }

    public Money totalAmount() {
        return subtotal().add(taxAmount());
    }

    public PaymentStatus paymentStatus() {
        return paymentStatus;
    }

    public Instant issuedAt() {
        return issuedAt;
    }
}

File: docs/article/source/go/apps/cargo-tracker/internal/billing/domain/invoice.go
package domain

import (
	"errors"
	"math"
	"strings"
	"time"

	shared "github.com/k2works/case-study-cargo-tracker/apps/cargo-tracker/internal/shared/domain"
)

// ドメインエラー。
var (
	ErrEmptyInvoiceId           = errors.New("invoice id must not be empty")
	ErrDiscountRateOutOfRange   = errors.New("discount rate must be between 0 and 0.30")
	ErrInvalidPaymentTransition = errors.New("invalid payment status transition")
)

// InvoiceId は請求書を一意に識別する値オブジェクト。
type InvoiceId struct{ value string }

// NewInvoiceId はバリデーション付きで InvoiceId を生成する。
func NewInvoiceId(v string) (InvoiceId, error) {
	if strings.TrimSpace(v) == "" {
		return InvoiceId{}, ErrEmptyInvoiceId
	}
	return InvoiceId{value: v}, nil
}

// Value は請求番号の文字列表現を返す。
func (i InvoiceId) Value() string { return i.value }

// BillingBookingId は Booking Context の予約を業務識別子で参照する値オブジェクト（BC 独立性）。
type BillingBookingId struct{ value string }

// NewBillingBookingId は予約参照 ID を生成する。
func NewBillingBookingId(v string) BillingBookingId { return BillingBookingId{value: v} }

// Value は予約 ID の文字列表現を返す。
func (b BillingBookingId) Value() string { return b.value }

// BillingShipperId は Shipper Context の荷主を業務識別子で参照し、法人判定を内包する値オブジェクト。
type BillingShipperId struct {
	shipperCode string
	shipperType string
}

// NewBillingShipperId は荷主参照 ID を生成する。
func NewBillingShipperId(shipperCode, shipperType string) BillingShipperId {
	return BillingShipperId{shipperCode: shipperCode, shipperType: shipperType}
}

// ShipperCode は荷主参照コードを返す。
func (s BillingShipperId) ShipperCode() string { return s.shipperCode }

// IsCorporate は法人荷主かを返す。
func (s BillingShipperId) IsCorporate() bool { return s.shipperType == "CORPORATE" }

// DiscountRate は割引率（0〜30%）を表す値オブジェクト。
type DiscountRate struct{ rate float64 }

// NewDiscountRate はバリデーション付きで割引率を生成する。
func NewDiscountRate(rate float64) (DiscountRate, error) {
	if rate < 0 || rate > 0.30 {
		return DiscountRate{}, ErrDiscountRateOutOfRange
	}
	return DiscountRate{rate: rate}, nil
}

// Value は割引率を返す。
func (d DiscountRate) Value() float64 { return d.rate }

// PaymentStatus は支払い状態を表す列挙型。
type PaymentStatus string

const (
	// PaymentStatusPending は未払い。
	PaymentStatusPending PaymentStatus = "PENDING"
	// PaymentStatusConfirmed は支払い確認済み（精算済み）。
	PaymentStatusConfirmed PaymentStatus = "CONFIRMED"
	// PaymentStatusOverdue は支払期限超過。
	PaymentStatusOverdue PaymentStatus = "OVERDUE"
	// PaymentStatusRefunded は返金済み。
	PaymentStatusRefunded PaymentStatus = "REFUNDED"
)

// Ja は支払い状態の日本語表示を返す。
func (s PaymentStatus) Ja() string {
	switch s {
	case PaymentStatusPending:
		return "未払い"
	case PaymentStatusConfirmed:
		return "精算済み"
	case PaymentStatusOverdue:
		return "期限超過"
	case PaymentStatusRefunded:
		return "返金済み"
	default:
		return string(s)
	}
}

// cargoTypeCoefficient は貨物種別の料金係数を返す（GENERAL 1.0 / HAZARDOUS 1.8 / REFRIGERATED 1.5）。
func cargoTypeCoefficient(t shared.CargoType) float64 {
	switch t {
	case shared.CargoTypeHazardous:
		return 1.8
	case shared.CargoTypeRefrigerated:
		return 1.5
	default:
		return 1.0
	}
}

// CalculateBaseAmount は基本料金を算出する。
//
//	基本料金 = 距離係数 × 重量（kg） × 貨物種別係数
//
// 距離係数（distanceFactor・円/kg 相当）は経路の実距離データが無いためスタブ値（区間数/所要日数
// ベースの簡易モデル）を合成ルート/サービスから渡す（IT8 注2）。
func CalculateBaseAmount(distanceFactor int64, weightKg float64, cargoType shared.CargoType) Money {
	amount := math.Round(float64(distanceFactor) * weightKg * cargoTypeCoefficient(cargoType))
	return NewMoney(int64(amount), "JPY")
}

// Invoice は精算書（請求書）を表す集約ルート。
type Invoice struct {
	invoiceId      InvoiceId
	bookingId      BillingBookingId
	shipperId      BillingShipperId
	baseAmount     Money
	discountRate   DiscountRate
	discountAmount Money
	taxAmount      Money
	finalAmount    Money
	paymentStatus  PaymentStatus
	issuedAt       time.Time
	dueDate        time.Time
	paidAt         time.Time
}

// GenerateInvoiceParams は請求書発行のパラメータ集合。
type GenerateInvoiceParams struct {
	InvoiceId    InvoiceId
	BookingId    BillingBookingId
	ShipperId    BillingShipperId
	BaseAmount   Money
	DiscountRate DiscountRate
	TaxRate      float64
	IssuedAt     time.Time
	DueDays      int
}

// GenerateInvoice は基本料金・割引率・税率から請求書を発行する（PENDING 状態・US21/US22）。
//
//	割引額 = 基本料金 × 割引率
//	割引後 = 基本料金 - 割引額
//	消費税 = 割引後 × 税率
//	合計   = 割引後 + 消費税
func GenerateInvoice(p GenerateInvoiceParams) (*Invoice, error) {
	discountAmount := p.BaseAmount.MultiplyRate(p.DiscountRate.Value())
	discounted, err := p.BaseAmount.Subtract(discountAmount)
	if err != nil {
		return nil, err
	}
	taxAmount := discounted.MultiplyRate(p.TaxRate)
	finalAmount, err := discounted.Add(taxAmount)
	if err != nil {
		return nil, err
	}
	return &Invoice{
		invoiceId:      p.InvoiceId,
		bookingId:      p.BookingId,
		shipperId:      p.ShipperId,
		baseAmount:     p.BaseAmount,
		discountRate:   p.DiscountRate,
		discountAmount: discountAmount,
		taxAmount:      taxAmount,
		finalAmount:    finalAmount,
		paymentStatus:  PaymentStatusPending,
		issuedAt:       p.IssuedAt,
		dueDate:        p.IssuedAt.AddDate(0, 0, p.DueDays),
	}, nil
}

// ConfirmPayment は入金を確認し精算済み（CONFIRMED）に遷移する（US23）。
func (i *Invoice) ConfirmPayment(paidAt time.Time) error {
	if i.paymentStatus != PaymentStatusPending && i.paymentStatus != PaymentStatusOverdue {
		return ErrInvalidPaymentTransition
	}
	i.paymentStatus = PaymentStatusConfirmed
	i.paidAt = paidAt
	return nil
}

// MarkOverdue は支払期限を超過した未払い請求を OVERDUE に遷移する（US23）。
// 支払期限は DATE 単位のため、時刻付きの now と日付単位で比較する
// （期限当日に時刻付きで判定しても OVERDUE にしない）。
func (i *Invoice) MarkOverdue(now time.Time) error {
	if i.paymentStatus != PaymentStatusPending {
		return nil
	}
	if dateOnly(now).After(dateOnly(i.dueDate)) {
		i.paymentStatus = PaymentStatusOverdue
	}
	return nil
}

// dateOnly は時刻を切り捨て日付（同一ロケーションの 00:00）に正規化する。
func dateOnly(t time.Time) time.Time {
	return time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, t.Location())
}

// ReconstructInvoice は永続化データから請求書を再構築する。
func ReconstructInvoice(p GenerateInvoiceParams, discountAmount, taxAmount, finalAmount Money, status PaymentStatus, paidAt time.Time) *Invoice {
	return &Invoice{
		invoiceId:      p.InvoiceId,
		bookingId:      p.BookingId,
		shipperId:      p.ShipperId,
		baseAmount:     p.BaseAmount,
		discountRate:   p.DiscountRate,
		discountAmount: discountAmount,
		taxAmount:      taxAmount,
		finalAmount:    finalAmount,
		paymentStatus:  status,
		issuedAt:       p.IssuedAt,
		dueDate:        p.IssuedAt.AddDate(0, 0, p.DueDays),
		paidAt:         paidAt,
	}
}

// アクセサ。
func (i *Invoice) InvoiceId() InvoiceId         { return i.invoiceId }
func (i *Invoice) BookingId() BillingBookingId  { return i.bookingId }
func (i *Invoice) ShipperId() BillingShipperId  { return i.shipperId }
func (i *Invoice) BaseAmount() Money            { return i.baseAmount }
func (i *Invoice) DiscountRate() DiscountRate   { return i.discountRate }
func (i *Invoice) DiscountAmount() Money        { return i.discountAmount }
func (i *Invoice) TaxAmount() Money             { return i.taxAmount }
func (i *Invoice) FinalAmount() Money           { return i.finalAmount }
func (i *Invoice) PaymentStatus() PaymentStatus { return i.paymentStatus }
func (i *Invoice) IssuedAt() time.Time          { return i.issuedAt }
func (i *Invoice) DueDate() time.Time           { return i.dueDate }
func (i *Invoice) PaidAt() time.Time            { return i.paidAt }
